plugins/chat/lib/chat/onebox_handler.rb

```diff
@@ -0,0 +1,75 @@
+# frozen_string_literal: true
+
+module Chat
+  class OneboxHandler
+    def self.handle(url, route)
+      if route[:message_id].present?
+        message = Chat::Message.find_by(id: route[:message_id])
+        return if !message
+
+        chat_channel = message.chat_channel
+        user = message.user
+        return if !chat_channel || !user
+      else
+        chat_channel = Chat::Channel.find_by(id: route[:channel_id])
+        return if !chat_channel
+      end
+
+      return if !Guardian.new.can_preview_chat_channel?(chat_channel)
+
+      args = {
+        url: url,
+        channel_id: chat_channel.id,
+        channel_name: chat_channel.name,
+        is_category: chat_channel.chatable_type == "Category",
+        color: chat_channel.chatable_type == "Category" ? chat_channel.chatable.color : nil,
+      }
+
+      if message.present?
+        render_message_onebox(args, message)
+      else
+        render_channel_onebox(args, chat_channel)
+      end
+    end
+
+    private
+
+    def self.render_message_onebox(args, message)
+      args[:message_id] = message.id
+      args[:username] = message.user.username
+      args[:avatar_url] = message.user.avatar_template_url.gsub("{size}", "20")
+      args[:cooked] = message.cooked
+      args[:created_at] = message.created_at
+      args[:created_at_str] = message.created_at.iso8601
+
+      Mustache.render(Chat.message_onebox_template, args)
+    end
+
+    def self.render_channel_onebox(args, chat_channel)
+      users =
+        chat_channel
+          .user_chat_channel_memberships
+          .includes(:user)
+          .where(user: User.activated.not_suspended.not_staged)
+          .limit(10)
+          .map do |membership|
+            {
+              username: membership.user.username,
+              avatar_url: membership.user.avatar_template_url.gsub("{size}", "60"),
+            }
+          end
```

Comment:
We have `Chat::ChannelMembershipsQuery` which might be used here instead I think